### PR TITLE
Add support for model metadata

### PR DESCRIPTION
**Pull request by mcslao**

### CODE REVIEW

File: client/python/samples/basic/JsonSample.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      "TYPE": "model_type",
+      "CONFIG": {
+        "l1": "20"
+      },
+      "METRICS": [
+        {"TYPE": "accuracy", "VALUE": 0.5},
+        {"TYPE" : "score", "VALUE": 0.9}
+      ],
+      "TAG": "train",
+      "extra-kv-pair": 2,
```
Comment (mpvartak) [inline client/python/samples/basic/JsonSample.json]: Change name?

File: client/python/samples/basic/YamlSample.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+---
```
Comment (mpvartak) [inline client/python/samples/basic/YamlSample.yaml]: Remove

File: frontend/public/js/m.js [hunk 3]
```diff
@@ -271,6 +271,10 @@ $(function() {
             obj[metadata[j].key] = metadata[j].value;
           }
 
+          console.log("Printing model metadata...");
```
Comment (mpvartak) [inline frontend/public/js/m.js]: Add comment to remove this after @weienlee's change is in.

File: frontend/routes/projects.js [hunk 4]
```diff
@@ -65,6 +65,7 @@ router.get('/:id/models', function(req, res, next) {
 router.get('/:id/ms', function(req, res, next) {
   var projectId = req.params.id;
   api.getProjectModels(projectId, function(response) {
+    console.log(response);
```
Comment (mpvartak) [inline frontend/routes/projects.js]: extraneous

File: pom.xml [hunk 5]
```diff
@@ -0,0 +1,150 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mpvartak) [inline pom.xml]: This file needs to go. From other branch.

**Review by mpvartak — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
This doesn't go so smoothly... gen_sqlite ran fine and mongod fired up. Running start_server.sh dumped out a bunch of errors:

✓ arubisov@lvappi01089 ~/modeldb/server $ ./start_server.sh
[INFO] Scanning for projects...
[WARNING]
[WARNING] Some problems were encountered while building the effective model for edu.mit.csail.db.ml:modeldb:jar:1.0-SNAPSHOT
[WARNING] 'build.plugins.plugin.version' for org.apache.maven.plugins:maven-compiler-plugin is missing. @ line 93, column 21
[WARNING]
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING]
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING]
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building modeldb 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.4.1:clean (default-clean) @ modeldb ---
[INFO] Deleting /home/arubisov/modeldb/server/target
[INFO]
[INFO] --- maven-resources-plugin:2.5:resources (default-resources) @ modeldb ---
[debug] execute contextualize
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] Copying 2 resources
[INFO]
[INFO] --- maven-compiler-plugin:2.3.2:compile (default-compile) @ modeldb ---
[WARNING] File encoding has not been set, using platform encoding UTF-8, i.e. build is platform dependent!
[INFO] Compiling 143 source files to /home/arubisov/modeldb/server/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/util/ExceptionWrapper.java:[4,14] error: package modeldb does not exist
[ERROR] /home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/storage/PipelineEventDao.java:[5,0] error: package modeldb does not exist
[ERROR] /home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/storage/PipelineEventDao.java:[32,78] error: cannot find symbol
[ERROR]  class PipelineEventDao
/home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/storage/PipelineEventDao.java:[32,21] error: cannot find symbol
[ERROR]  class PipelineEventDao
/home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/storage/PipelineEventDao.java:[47,44] error: cannot find symbol
[ERROR]  class PipelineEventDao
/home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/storage/PipelineEventDao.java:[47,16] error: cannot find symbol
[ERROR]  class PipelineEventDao
/home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/algorithm/similarity/SimilarModels.java:[7,14] error: package modeldb does not exist
[ERROR] /home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/algorithm/similarity/SimilarModels.java:[8,14] error: package modeldb does not exist
[ERROR] /home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/algorithm/similarity/SimilarModels.java:[9,14] error: package modeldb does not exist
[ERROR] /home/arubisov/modeldb/server/src/main/java/edu/mit/csail/db/ml/server/algorithm/similarity/SimilarModels.java:[29,57] error: cannot find symbol
[ERROR]  class SimilarModels
...etc

**Comment by mpvartak** [comment]
Hi @arubisov,

This has to do with the thrift version (it can't find modeldb package). Note that the build process has changed slightly. If your thrift version is x.y.z, you would run ./start_server.sh x.y.z

This will fix the error.

Manasi

**Review by mpvartak — APPROVED** [review]

File: client/python/modeldb/basic/ModelDbSyncerBase.py [hunk 6]
```diff
@@ -371,4 +373,46 @@ def get_dataset_for_tag(self, data_tag):
             data_tag = "default"
         return self.datasets[data_tag]
 
-    # TODO: do we want a sync all?
+    def dataset_from_dict(self, dataset_dict):
+        filename = dataset_dict[metadata_constants.DATASET_FILENAME_KEY]
+        metadata = dataset_dict[metadata_constants.DATASET_METADATA_KEY] \
+            if metadata_constants.DATASET_METADATA_KEY in dataset_dict else {}
```
Comment (geffy) [inline client/python/modeldb/basic/ModelDbSyncerBase.py]: Can be done shortly: `dataset_dict.get(metadata_constants.DATASET_METADATA_KEY, {})`

Comment (mpvartak) [reply client/python/modeldb/basic/ModelDbSyncerBase.py]: great, fixed!

**Pull request merged**